src/Request.php

```diff
@@ -117,6 +117,19 @@ public function withUri(UriInterface $uri, $preserveHost = false)
         return $new;
     }
 
+    /**
+     * Create a PSR7 Request instance from another PSR7 Request instance.
+     *
+     * This is intended to be used to convert already-existing Request instances
+     * to instances of a subclass which may have different capabilities.
+     *
+     * @param RequestInterface
+     * @return static
+     */
+    public static function fromPsr7(RequestInterface $src) {
+        return new static($src->getMethod(), $src->getUri(), $src->getHeaders(), $src->getBody(), $src->getProtocolVersion());
```

Comment:
I believe this solution has the exact same flaws as discussed in the referenced PRs/issues: there is no contract on the constructor, so if a class extending this request has a different constructor signature, it will simply break.

One possible solution would ironically be making the contract more strict (eg. making the constructor final), but that would be a BC break.

Honestly I really don't see what's wrong with implementing a wrapper class or some other perfectly fine solution. The only contract you need to keep is the interface, beyond that you do whatever you want, you are not tied to this library at all.

If you need some custom behavior, you are free to implement it, and thanks to the PSR it's even super easy to do.

My opinion on the matter is generic libraries can't provide support for all use cases, only a number of them, but since we build our softwares on contracts mostly, not implementations, this should not be a problem.

Reply:
Found @Tobion 's comment about suggesting this method. As he is the maintainer of this package I'm okay with it if he accepts, but my opinion is still up.

Reply:
Looking back over these conversations, my gut reaction is to wonder why on Earth we're trying so hard to police what other people are doing with our open source code. Most of our discussions have centered around, "What if someone extends our class in a way that breaks their code??" My opinion (which you are free to reject) is that that's their right and their issue, not ours. Furthermore, the idea that someone might do something with our code that we never even dreamed of is a feature, not a bug.

In short, I think we need to be careful about limiting people's possibilities to the boundaries of our own imaginations. The great value of openness lies precisely in all the things people do with our code that we never thought to do ourselves.

Now, to the issue at hand, you could argue that you don't want to muddy up your codebase with everyone's extraneous features and that's why you don't want to let any of my suggestions in. Really, though, you'd only be making my point for me: I'm trying to introduce the minimal code necessary to allow me and others to _build on_ the work that you've already done without having to PR everything into your codebase or argue with you about what the definition of a fundamental Request or Response is.

Yes, I could build an adapter. In fact, I did end up building an adapter, because I realized you'd never merge my PR. That took me more than an hour with all the necessary tests. Was that a good use of my time? Absolutely not. And I wish you would work with me here to merge this tiny addition to your codebase so that no one ever has to spend another hour doing something that should take only 5 minutes, and so that you don't ever have to spend another minute bikeshedding about what features should and shouldn't be in your package.

Love, Kael